lib/src/working_copy.rs

```diff
@@ -676,85 +694,93 @@ impl TreeState {
             },
         );
         struct WorkItem {
-            dir: RepoPath,
-            disk_dir: PathBuf,
+            path: RepoPath,
+            disk_path: PathBuf,
+            disk_file_type: DiskFileType,
             git_ignore: Arc<GitIgnoreFile>,
         }
```

Comment:
> This is useful for filesystem monitors because they would like to return a list of changed files rather than a list of changed directories.

I don't follow. The current Matcher-based approach seems sufficient to me. I thought we just needed to skip updating the `file_states` when using Watchman. Why doesn't that work?

Reply:
@martinvonz The original justification for this commit was so that we could populate the initial `work` list with only files from Watchman, instead of starting from the root directory. However, doing so means that we don't properly process the `.gitignore`s from ancestor directories.

In the latest version, we don't populate `work` directly with the files from Watchman. Instead, we set up the `Matcher`s in such a way that we start traversal from the root directory and only visit paths that end up containing Watchman-changed-files. This means that we query for the contents of each ancestor directory of a changed file, even though we know exactly which files we want to visit.

In principle, this commit would let us skip that traversal by starting with our own `work` list (and manually looking for `.gitignore`s in ancestor directories, which should still be faster than querying for the directory contents for each ancestor directory?), but that's no longer in this commit.

I can drop this commit if you like, or we can keep it in the case that we want to make that optimization one day. (I also personally think it's cleaner to operate on arbitrary files/directories/symlinks rather than require starting from a containing directory, but I suppose that's a matter of wisdom and I might be wrong about the usefulness in this case.)